Add PortAudio output support

kareandersen opened the pull request:
Here is code to support output over PortAudio. It should allow users to listen to SNDH tracks in MacOS without having to deal with Core Audio. As a bonus, it should work on any operating system that has a PortAudio port, including Linux and even MS Windows... Hope you find it useful to keep upstream - please don't hesitate to ask for any changes!

frno7 reviewed (commented):
Thanks, @kareandersen! Sounds good, but please review enclosed comments. Also, it seems Port Audio is unable to resume from ^Z (suspend), at least on Linux? Maybe also on Mac? Is it possible to fix, as suspend/resume works with ALSA? Finally, I suppose we should mention the new `PORTAUDIO` output option in `README.md`, `INSTALL`, and the `doc/psgplay.1` manual page?

frno7 commented:
By the way, what Mac OS variant have you tested it on? My Mac is unable to install anything later than [High Sierra](https://en.wikipedia.org/wiki/MacOS_High_Sierra), a model somewhat out of fashion these days. Oh, the workflow script seems to have problems with slashes in the branch names (`psgplay-web-29/merge-af8fc3b0`). I’ll look into it for a fix.

kareandersen commented on lib/out/portaudio.c:
```diff
@@ -0,0 +1,228 @@
+
+static void *portaudio_open(const char *output, int frequency, bool nonblocking)
+{
+	PaStreamParameters params;
```
Thanks! Added (my C is rusty!)

kareandersen commented on lib/out/portaudio.c:
```diff
@@ -0,0 +1,228 @@
+		Pa_GetDeviceInfo(params.device)->defaultLowOutputLatency;
+	params.hostApiSpecificStreamInfo = NULL;
+
+	err = Pa_OpenStream(&stream, NULL, &params, frequency, 4096, paClipOff, NULL, NULL);
```
Thanks again, absolutely. Fixed.

kareandersen commented on system/unix/Makefile:
```diff
@@ -52,6 +52,14 @@ PSGPLAY_ALSA_LIB := $(shell pkg-config --silence-errors --libs alsa || echo -las
 
+ifeq (1,$(PORTAUDIO))
+PSGPLAY_PORTAUDIO_LIB := $(shell pkg-config --silence-errors --libs portaudio-2.0)
+PSGPLAY_PORTAUDIO_CFLAGS := $(shell pkg-config --silence-errors --cflags portaudio-2.0)
```
Good catch, fixed. I think we should not add any fallback to these as they are not necessarily expected to be found on a typical setup unlike ALSA is on Linux. Better indeed to handle it through pkg-config.

kareandersen commented:
> By the way, what Mac OS variant have you tested it on? My Mac is unable to install anything later than [High Sierra](https://en.wikipedia.org/wiki/MacOS_High_Sierra), a model somewhat out of fashion these days. Oh, the workflow script seems to have problems with slashes in the branch names (`psgplay-web-29/merge-af8fc3b0`). I’ll look into it for a fix.

Tested on my MacOS 15.3 Sequoia MBP (M1 CPU). Most days I run Linux on it (Asahi/Fedora), but sometimes you just have to roll up your sleeves and support the community... :)

I guess we could add Actions for these sorts of builds too if you are interested?

Also working on a separate PR to support dylib building on MacOS (no elf/.so support here - so the build craps out). And on getting the trace/disassembler up by building with homebrew gcc instead of clang (which does not support the endian attributes). Perhaps we should discuss that in a scope of its own...

kareandersen commented:
> <snip>Also, it seems Port Audio is unable to resume from ^Z (suspend), at least on Linux? Maybe also on Mac? Is it possible to fix, as suspend/resume works with ALSA? </snip>

Hmm suspend works here, using Kitty terminal with tmux even. I'll give it a spin on "a real os" later to see if I can reproduce it and/or figure it out.

frno7 commented:
Many thanks @kareandersen! Merged in commit ca3338bc99efb9466e8a3ba7ac951d17b65ab0e3, with minor editing. For example, I removed _either_ in “either ALSA or PortAudio”, since these options are not actually mutually exclusive. Both can happily coexist, in which case PortAudio will be defaulted, but ALSA can be selected with for instance `-o alsa:default`. :-)

I (hopefully) also fixed the slash problems in the workflow script, in commit b8378dfc8d5c9d7133fdd6c33ad0aa5aa1c9141a. To be tested with forthcoming pull requests.

Regarding the suspend/resume problem: Curiously, if I suspend/resume twice, that is suspend-resume-suspend-resume, it works: PortAudio will resume playing, suggesting there’s some internal PortAudio state toggle malfunction. Also,
the problem only manifests itself with interactive text mode (the TUI). Command mode seems unaffected. There is some special handling here involving `sample_buffer_pause` and `sample_buffer_resume`:

https://github.com/frno7/psgplay/blob/ca3338bc99efb9466e8a3ba7ac951d17b65ab0e3/system/unix/text-mode.c#L241-L267

kareandersen commented:
> Many thanks @kareandersen! Merged in commit [ca3338b](https://github.com/frno7/psgplay/commit/ca3338bc99efb9466e8a3ba7ac951d17b65ab0e3), with minor editing. For example, I removed _either_ in “either ALSA or PortAudio”, since these options are not actually mutually exclusive. Both can happily coexist, in which case PortAudio will be defaulted, but ALSA can be selected with for instance `-o alsa:default`. :-)
> 

Ah, excellent, I had not considered that path :)

> I (hopefully) also fixed the slash problems in the workflow script, in commit [b8378df](https://github.com/frno7/psgplay/commit/b8378dfc8d5c9d7133fdd6c33ad0aa5aa1c9141a). To be tested with forthcoming pull requests.

Thank you!

> Regarding the suspend/resume problem: Curiously, if I suspend/resume twice, that is suspend-resume-suspend-resume, it works: PortAudio will resume playing, suggesting there’s some internal PortAudio state toggle malfunction. Also, the problem only manifests itself with interactive text mode (the TUI). Command mode seems unaffected. There is some special handling here involving `sample_buffer_pause` and `sample_buffer_resume`:
> 
> https://github.com/frno7/psgplay/blob/ca3338bc99efb9466e8a3ba7ac951d17b65ab0e3/system/unix/text-mode.c#L241-L267

I just got back here now to report the same thing under Linux, and am looking into ways of handling it. I don't like things like this hanging around, possibly hiding a larger issue. I also noticed that it will play the buffer remnants after closing, which is not ideal behavior - if I shut it down I would like instant silence. Let's shake the tree a bit and see what falls out...

Anyway, thanks for accepting. This is an awesome project for sure! More hackery to follow...

frno7 commented:
> > Many thanks @kareandersen! Merged in commit [ca3338b](https://github.com/frno7/psgplay/commit/ca3338bc99efb9466e8a3ba7ac951d17b65ab0e3), with minor editing. For example, I removed _either_ in “either ALSA or PortAudio”, since these options are not actually mutually exclusive. Both can happily coexist, in which case PortAudio will be defaulted, but ALSA can be selected with for instance `-o alsa:default`. :-)
> 
> Ah, excellent, I had not considered that path :)

In the future, we might want to accept `-o portaudio:` as a means to explicitly select PortAudio. I don’t think it’s particularly relevant now, but in case even more output options are added, it might become necessary.

> Anyway, thanks for accepting. This is an awesome project for sure! More hackery to follow...

You’re welcome!
